app/services/messages/channels.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+module Messages
+  class Channels < Base
+    attr_accessor :is_private
+
+    def initialize(params)
+      @is_private = ActiveRecord::Type::Boolean.new.type_cast_from_database(params[:is_private])
+      super(params)
+    end
+
+    def related_users
+      case app_name
+      when DONOR_APP, ADMIN_APP
+        admin_donor_mentions.uniq
+      when STOCK_APP, BROWSE_APP
+        stock_browse_mentions.uniq
+      end
+    end
+
+    def admin_donor_mentions
+      if is_private
+        get_users_by_roles(['Supervisor'])
+      else
+        admin_donor_channel_users
+      end
+    end
+
+    def admin_donor_channel_users
+      users = get_users_by_roles(%w[Supervisor Reviewer]).uniq
+      users << { id: messageable.created_by_id, name: messageable.created_by.full_name } unless owner?
+      users
+    end
+
+    def stock_browse_mentions
+      users = get_users_by_roles(['Order fulfilment', 'Order administrator'])
+      users << { id: messageable.created_by_id, name: messageable.created_by.full_name } unless owner?
+      users
+    end
+
+    private
+
+    def user_roles
+      User.where.not(users: { id: current_user.id, disabled: true })
+          .joins(:user_roles)
+          .joins(:roles)
+    end
+
+    def get_users_by_roles(roles)
+      user_roles.where(roles: { name: roles }).uniq
+                .map { |user| { id: user.id, name: user.full_name } }
+    end
+  end
+end
```

Comment:
As we discussed offline, we'll only have the @ mention functionality for Admin and Stock apps. therefore this class can be massively simplified. In fact, I wonder if it's best to handle the id, full_name in a new serializer and the decisions about which user roles to include can be made using params and scopes in the controller... here's what I mean:

```
def UsersController
    def mentionable_users
        ALLOWED_ROLES = ['Reviewer', 'Supervisor', 'Order Administrator', 'Order Fulfilment']
        roles = (ALLOWED_ROLES & params['roles'].split(',').map(&:strip).uniq) # allow only certain roles
        @users = @users.not_disabled
          .not_deleted
          .not_current_user(current_user.id)
          .with_roles(roles)
          render json: @users, each_serializer: UserMentionableSerializer
      end
end

class User
  scope  :not_disabled, ->{ where(is_disabled: false) }
  scope :with_roles, ->(role_names) { roles.where(name: role_names) }
end

class UserMentionableSerializer < ApplicationSerializer
  embed :ids
  attributes :id, :full_name
end
```

Doing this would mean you can delete the `messages/channels.rb` class